python/dex/interop/jax.py

```diff
@@ -271,5 +271,127 @@ def tuple_string(prefix):
 
 jax.interpreters.ad.primitive_jvps[dex_call_p] = dex_call_jvp
 
-# TODO
-# jax.interpreters.ad.primitive_transposes[self.primitive] = ...
+# === transpose ===
+
+# alias to avoid confusion around overloading of "primal".
+_is_linear_input = jax.ad.is_undefined_primal
+
+
+def dex_call_evaluate_linearized_transpose(cotangents, *args, func_atom):
+  """Evaluates the transpose of a function atom.  """
+
+  # `func_atom` is assumed to be of the form of `evaluate_linearized` from
+  # `dex_call_jvp`, applied to a some function atom, called `f`, say.
+  # Concretely, if `f` has three primal arguments, `func_atom` should look like:
+  # ```
+  # \ x0 x1 x2 u0 u1 u2.
+  #   intermediate_linearized = linearize f (x0, x1, x2)
+  #   snd intermediate_linearized (u0 u1 u2)
+  # ```
+  # In particular, its arguments are assumed to be `num_primals` primal inputs,
+  # followed by `num_primals` tangent inputs.
+
+  assert len(args) % 2 == 0
+  num_primals = len(args) // 2
+
+  primals, tangents = args[:num_primals], args[num_primals:]
+
+  # Helper functions to build strings of primal and tangent inputs.
+  def arg_string(prefix, index_set):
+    return " ".join(f"{prefix}{i}" for i in index_set)
+
+  def tuple_string(prefix, index_set):
+    return "(" + ", ".join(f"{prefix}{i}" for i in index_set) + ")"
+
+  # JAX uses `UndefinedPrimal` instances to mark input variables which the
+  # function needs to be transposed with respect to, and (consequently) for
+  # which no concrete values are available. `_is_linear_input` tests if the
+  # input is such an instance.
+  #
+  # `func_atom` is only guaranteed to be linear in its tangent inputs, so we
+  # check here that we're not expected to tranpose it with respect to any primal
+  # inputs. JAX *should* take care of this automatically, but this mechanism is
+  # somewhat poorly documented so its worth double checking.
+  if any(_is_linear_input(p) for p in primals):
+    raise RuntimeError("Primal inputs to transpose primitive are undefined.")
+
+  # Add `func_atom` as a variable `linearized` in the context.
+  env = api.insert(func_atom.module, api.as_cstr("linearized"), func_atom)
+
+  # Form lists of the indices in `tangents` which correspond to linear inputs
+  # (which we are expected to transpose w.r.t.) and constant inputs (which we
+  # are not).
+  tangent_input_indices = [
+      i for i, t in enumerate(tangents) if _is_linear_input(t)
+  ]
+  tangent_constant_indices = [
```

Comment:
Is that ever non-empty? I don't think so? Or when you say "constant inputs" do you mean the zeros that you've instantiated in the JVP? If yes then let's make the comment more precise.

Reply:
It's the latter, added a further comment to clarify.

The linear / constant input terminology I was using comes from [here](https://jax.readthedocs.io/en/latest/notebooks/How_JAX_primitives_work.html#transposition), if there's a better way to state this please feel free to suggest something!

Reply:
Looks great, thanks!